lib/spack/spack/spec.py

```diff
@@ -2095,31 +2096,107 @@ def traverse_edges(
+
             edge_attributes = ""
-            if item.virtuals or item.when != Spec():
-                edge_attributes = item.spec.format("{edge_attributes}") + " "
+            if item.when != Spec():
+                edge_attributes = self._format_edge_attributes(
+                    item, deptypes=deptypes, virtuals=False
+                )
+                if edge_attributes:
+                    edge_attributes += " "
+            virtuals = f"{','.join(item.virtuals)}=" if item.virtuals else ""
+
+            parts.append(
+                f"%{edge_attributes}{virtuals}{item.spec.format(format_string)}".replace(
+                    current_name, new_name
+                )
+            )
 
-            parts.append(f"%{edge_attributes}{item.spec.format()}".replace(current_name, new_name))
         for item in sorted(transitive, key=lambda x: x.spec.name):
-            # Recurse to attach build deps in order
+            if not predicate(item):
+                continue
+
             edge_attributes = ""
-            if item.virtuals or item.depflag or item.when != Spec():
-                edge_attributes = item.spec.format("{edge_attributes}") + " "
-            parts.append(f"^{edge_attributes}{str(item.spec)}")
+            if item.depflag or item.when != Spec():
+                edge_attributes = self._format_edge_attributes(
+                    item, deptypes=deptypes, virtuals=False
+                )
+                if edge_attributes:
+                    edge_attributes += " "
+
+            virtuals = f"{','.join(item.virtuals)}=" if item.virtuals else ""
+
+            sigil = "%" if _force_direct else "^"  # hack until direct deps are represented better
+            parts.append(f"{sigil}{edge_attributes}{virtuals}{item.spec.format(format_string)}")
+
+            # also recursively add any build dependencies of transitive dependencies
+            if item.spec._dependencies:
+                parts.append(
+                    item.spec._format_dependencies(
+                        format_string=format_string,
+                        predicate=predicate,
+                        deptypes=deptypes,
+                        _force_direct=_force_direct,
+                    )
+                )
+
         return " ".join(parts).strip()
 
+    @property
+    def compilers(self):
+        # TODO: get rid of the space here and make formatting smarter
+        return " " + self._format_dependencies(
+            "{name}{@version}",
+            predicate=lambda dep: any(lang in dep.virtuals for lang in ("c", "cxx", "fortran")),
+            deptypes=False,
+            _force_direct=True,
+        )
+
+    @property
+    def long_spec(self):
+        """Returns a string of the spec with the dependencies completely enumerated."""
+        string = self.format()
+        deps = self._format_dependencies()
+        string = f"{string} {deps}" if deps else string
```

Comment:
If you're going to strip the string, this can just be
```suggestion
        string = f"{string} {deps}"
```

Reply:
simplified